### PR TITLE
a unified data structure for all kinds of data

### PR DESCRIPTION
**Pull request by L1aoXingyu**
## 问题描述

因为不同类型的数据集会有不同的输入，所以在 data 传到 `nn.Module` 或者 `nn.Graph` 之前，需要一个 `get_batch` 函数做处理，比如在 `tools/pretrain_bert.py` 下面，需要这样一段代码进行处理

https://github.com/Oneflow-Inc/libai/blob/e730623acb9a5ab4c33cf96cd2f5aef2dbd2d5e3/tools/pretrain_bert.py#L28-L50

假如换一个模型，除了这个 `get_batch` 之外，其他的所有代码都是相同的，但是需要新写一个 python 文件，这样不仅增加了很多冗余代码，减少代码可读性，而且只要换一个数据集，或者输入类型，就需要重写一个类似 `pretrain_bert.py` 文件，会让用户体验变差。

## 解决方案

可以考虑添加一个默认的 `get_batch` 方法到 trainer 上，主要的职责就是将 local tensor 转移到 consistent tensor 上，只需要保证 `__getitem__` 返回的结果顺序和 `nn.Module` 或者 `nn.Graph` 传参顺序一致就可以了，这样相当于用户重写一个 dataset 只需要保证和网络输入顺序一致，那么便可以使用默认的 `get_batch` 代码，相当于用户可以直接理解为 dataset 的 `__getitem__` 和模型对接，中间的过程由 libai 内置的函数去接管。

如果没有流水并行，这样做已经足够了，但是在流水并行下，需要输入网络部分的 tensor 在第一阶段的 placement 上，算 loss 部分 (比如 label) 的 tensor 在最后一个阶段的 placement 上，这时默认的 `get_batch` 并不知道哪些 tensor 是输入相关，哪些是 loss 相关，所以无法自动做到这一点。

这里的 sbp 和 placement 需要交给用户去指定，因为写 dataset 的人一定知道哪些和输入有关，哪些和 loss 有关，所以可以在 `__getitem__` 的时候对 sbp 和 placement 进行指定，但是在这里并不能直接将 local tensor 转成 consistent tensor，因为这只是单个样本，所以需要返回一个类似于 (tensor, sbp, placement) 的数据结构，在获得一个 batch 数据之后，最终在 `get_batch` 阶段再延后进行 to_consistent 的操作。

所以可以考虑设计一个数据结构去存放这些信息，同时考虑到 libai 中的数据包含着图片，文本等多模态信息，那么指定一种统一的数据类型可以让后续的工作变得更方便，相比于搞一个 dict 什么都往里塞，多了一点抽象上的方便，同时也有 research 需要的灵活性。

处理我们自定的数据结构只需要在构建 DataLoader 中传入一个重写的 `collate_fn`

```python
def trivial_batch_collator(batch):
    assert isinstance(
        batch[0], Instance
    ), "batch[0] must be `instance` for trivial batch collator"
    batch = Instance.stack(batch)

    return batch
```

最后可以将 `pretrain_bert.py` 改成 `train_net.py`，这个训练的 interface 基本可以支持大多数的常规训练需求，就不需要换一个数据集，换一个模型就重写一个 `train_net.py`.

### CODE REVIEW

File: libai/data/structures.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+from dataclasses import dataclass, field
+from collections import OrderedDict
+from typing import List
+import oneflow as flow
+
+from libai.utils import distributed as dist
+
+
+@dataclass
+class Metadata:
```
Comment (L1aoXingyu) [inline libai/data/structures.py:25]: 这个类的作用是为了存放类似 (tensor, sbp, placement) 这些内容和属性。

File: libai/data/structures.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    tensor: flow.Tensor
+    sbp_list: list = field(default_factory=lambda: ["split 0", "broadcast"])
```
Comment (L1aoXingyu) [inline libai/data/structures.py:27]: 为什么不适用 `flow.sbp.split(0)` 这样实际的 sbp 而是使用一个字符串，是因为 `flow.sbp.sbp` 在 dataloader 中不能进行 pickle.dump

File: libai/data/structures.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    2. ``len(instance)`` returns the number of instance
+    """
+
+    def __init__(self, **kwargs):
+
+        self._fields = OrderedDict()
+        for k, v in kwargs.items():
+            self.set(k, v)
+
+    def __setattr__(self, name, val):
```
Comment (rentainhe) [inline libai/data/structures.py:110]: 感觉可以补充完善一下传参类型提示
```python
def __setattr__(self, name: str, val: Any) -> None:
    pass
```

File: libai/data/structures.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+        if placement is not None:
+            self.placement = placement
+        else:
+            self.placement = dist.get_layer_placement(self.placement_idx)
+
+        self.tensor = self.tensor.to_consistent(sbp=self.sbp, placement=self.placement)
+
+    @staticmethod
+    def stack(metadata_lists):
```
Comment (rentainhe) [inline libai/data/structures.py:56]:
```python
@staticmethod
def stack(metadata_lists: ["Metadata"]) -> "Metadata":
    pass
```
- 感觉加一个提示信息会好些，以免后续还要继续开PR加

File: libai/data/structures.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        assert len(metadata_lists) > 0
+        if len(metadata_lists) == 1:
+            metadata_lists[0].tensor.unsqueeze_(0)  # add batch dim
+            return metadata_lists[0]
+
+        data0 = metadata_lists[0]
```
Comment (rentainhe) [inline libai/data/structures.py:62]: 这里可不可以写成
```python
tensor_size = metadata_lists[0].size()
sbp_list = metadata_lists[0].sbp_list
placement_idx = metadata_lists[0].placement_idx

for metadata in metadata_lists[1:]:
    assert metadata.size() == tensor_size, "info"
    assert metadata.sbp_list == sbp_list, "info"
    assert metadata.placement_idx == placement_idx, "info"
```

File: libai/trainer/default.py [hunk 6]
```diff
@@ -343,9 +343,27 @@ def train(self):
...
         super().train(self.start_iter, self.max_iter)
 
-    def run_step(self, get_batch: Callable):
+    def run_step(self):
         self._trainer.iter = self.iter
-        self._trainer.run_step(get_batch)
+        self._trainer.run_step(self.get_batch)
+
+    @classmethod
+    def get_batch(cls, data: Instance):
```
Comment (CPFLAME) [inline libai/trainer/default.py:351]: 我们是在哪一步, 把dataloader的dict输出,  转成Instance的?

Comment (L1aoXingyu) [reply libai/trainer/default.py:351]: 这个我写在上面去吧，忘了写流程了
其实我们是在 __getitem__ 里面定义好 instance，然后就不用管了，后续都可以自动做完所有的处理

File: libai/data/structures.py (hunk 1 above)
Comment (CPFLAME) [reply libai/data/structures.py:25]: 这里叫ConsistentTensorData 我感觉更准确一点,   但是这个名字有点长...

Comment (CPFLAME) [reply libai/data/structures.py:25]: ConsistentTensorData用户不一定理解,  不如叫DistTensorData?

**Comment by L1aoXingyu** [comment]
### 整体流程

1. 定义 dataset 进行单个数据的处理流程，比如

```python
class BertDataset(Dataset):
   
  def __getitem__(self, idx):
    # ...
    return Instance(
    	tokens=Metadata(tensor=..., placement_idx=0), # default sbp
      masks=Metadata(tensor=..., sbp_list=["split_0", "broadcast"]), # default placement
      labels=Metadata(tensor=..., placement_idx=-1)
    )
```
其中在 getitem 里需要返回一个 instance 去保存这个示例的属性，也就是网络需要的输入输出，定义顺序和 key 需要和 nn.Graph 中 build 的输入一样，这样即可以完成自动的 input unbind。

这里为了处理 consistent tensor 的问题，使用了 metadata，但是需要注意的时候，instance 并不是只支持接受 metadata 类型，他也可以支持一般的 numpy array，local tensor， list 等等常见的对象，这给复杂任务的处理提供了灵活性。

2. 通过 dataloader + sampler 的方式，取对应的 index，然后获得了一个 instance list，这个 list 表示的是一个 batch 的输入，通过 dataloader 中定义的 collate_fn 对结果进行 stack，获得一个 batch 的输入；
3. 在 DefaultTrainer 里面定义的 get_batch 函数对结果进行 to_consistent 的操作，即将 instance 中 metadata 里面的 local tensor `to_consistent` 到对应的 sbp 和 placement 上，然后输入到网络中；

通过这种方式，如果只是增加一个数据集，或者是网络接受不同的输入时，只需要在定义 dataset 的 `__getitem__` 时指定好对应的 sbp 和 placement_idx，而大部分情况都只需要指定 placement_idx，sbp 可以使用默认配置。

而如果用户需要比较复杂的操作，比如 mixup，在拿到 batch 数据之后需要对 data 和 label 进行混合，那么可以在 get_batch 中进行处理，比如只需要在 local tensor 上做 mixup，那么可以在 to_consistent 之前做 mixup，如果想在全局上做 mixup，那么可以在 to_consistent 之后做，因为我们将 get_batch 的接口绑定在 DefaultTrainer 上，所以用户只要重载这个方法就可以了。

File: libai/data/structures.py (hunk 1 above)
Comment (Ldpe2G) [reply libai/data/structures.py:25]: 我觉得 Dist 更容易理解

**Review by CPFLAME — APPROVED** [review]
